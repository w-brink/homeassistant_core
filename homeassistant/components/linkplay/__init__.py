"""Support for LinkPlay devices."""

from linkplay.bridge import LinkPlayBridge
from linkplay.discovery import linkplay_factory_bridge

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_HOST
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryNotReady
from homeassistant.helpers.aiohttp_client import async_get_clientsession

from .const import PLATFORMS


class LinkPlayData:
    """Data for LinkPlay."""

    bridge: LinkPlayBridge


type LinkPlayConfigEntry = ConfigEntry[LinkPlayData]


async def async_setup_entry(hass: HomeAssistant, entry: LinkPlayConfigEntry) -> bool:
    """Async setup hass config entry. Called when an entry has been setup."""

    session = async_get_clientsession(hass)
    if (
        bridge := await linkplay_factory_bridge(entry.data[CONF_HOST], session)
    ) is None:
        raise ConfigEntryNotReady(
            f"Failed to connect to LinkPlay device at {entry.data[CONF_HOST]}"
        )

    entry.runtime_data = LinkPlayData()
    entry.runtime_data.bridge = bridge
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
    return True


async def async_unload_entry(hass: HomeAssistant, entry: LinkPlayConfigEntry) -> bool:
    """Unload a config entry."""

    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
