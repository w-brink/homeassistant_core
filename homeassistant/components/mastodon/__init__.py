"""The Mastodon integration."""

from __future__ import annotations

from mastodon.Mastodon import Mastodon, MastodonError

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import (
    CONF_ACCESS_TOKEN,
    CONF_CLIENT_ID,
    CONF_CLIENT_SECRET,
    CONF_NAME,
    Platform,
)
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryNotReady
from homeassistant.helpers import discovery

from .const import CONF_BASE_URL, DOMAIN
from .utils import create_mastodon_client


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up Mastodon from a config entry."""

    try:
        client, _, _ = await hass.async_add_executor_job(
            setup_mastodon,
            entry,
        )

    except MastodonError as ex:
        raise ConfigEntryNotReady("Failed to connect") from ex

    assert entry.unique_id

    await discovery.async_load_platform(
        hass,
        Platform.NOTIFY,
        DOMAIN,
        {CONF_NAME: entry.title, "client": client},
        {},
    )

    return True


def setup_mastodon(entry: ConfigEntry) -> tuple[Mastodon, dict, dict]:
    """Get mastodon details."""
    client = create_mastodon_client(
        entry.data[CONF_BASE_URL],
        entry.data[CONF_CLIENT_ID],
        entry.data[CONF_CLIENT_SECRET],
        entry.data[CONF_ACCESS_TOKEN],
    )

    instance = client.instance()
    account = client.account_verify_credentials()

    return client, instance, account
