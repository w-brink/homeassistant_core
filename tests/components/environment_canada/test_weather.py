"""Test weather."""

import json

from syrupy.assertion import SnapshotAssertion

from homeassistant.components.weather import (
    DOMAIN as WEATHER_DOMAIN,
    SERVICE_GET_FORECASTS,
)
from homeassistant.core import HomeAssistant

from . import init_integration

from tests.common import load_fixture


async def test_forecast_daily(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
) -> None:
    """Test basic forecast."""

    ec_data = json.loads(
        load_fixture("environment_canada/current_conditions_data.json")
    )

    # First entry in test data is a half day; we don't want that for this test
    del ec_data["daily_forecasts"][0]

    await init_integration(hass, ec_data)

    response = await hass.services.async_call(
        WEATHER_DOMAIN,
        SERVICE_GET_FORECASTS,
        {
            "entity_id": "weather.home_forecast",
            "type": "daily",
        },
        blocking=True,
        return_response=True,
    )
    assert response == snapshot


async def test_forecast_daily_with_some_previous_days_data(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
) -> None:
    """Test forecast with half day at start."""

    ec_data = json.loads(
        load_fixture("environment_canada/current_conditions_data.json")
    )

    await init_integration(hass, ec_data)

    response = await hass.services.async_call(
        WEATHER_DOMAIN,
        SERVICE_GET_FORECASTS,
        {
            "entity_id": "weather.home_forecast",
            "type": "daily",
        },
        blocking=True,
        return_response=True,
    )
    assert response == snapshot
