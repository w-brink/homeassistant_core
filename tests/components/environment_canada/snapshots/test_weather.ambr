# serializer version: 1
# name: test_forecast_daily
  dict({
    'weather.home_forecast': dict({
      'forecast': list([
        dict({
          'condition': 'sunny',
          'datetime': '2022-10-04 15:00:00+00:00',
          'precipitation_probability': 0,
          'temperature': 18.0,
          'templow': 3.0,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2022-10-05 15:00:00+00:00',
          'precipitation_probability': 0,
          'temperature': 20.0,
          'templow': 9.0,
        }),
        dict({
          'condition': 'partlycloudy',
          'datetime': '2022-10-06 15:00:00+00:00',
          'precipitation_probability': 0,
          'temperature': 20.0,
          'templow': 7.0,
        }),
        dict({
          'condition': 'rainy',
          'datetime': '2022-10-07 15:00:00+00:00',
          'precipitation_probability': 40,
          'temperature': 13.0,
          'templow': 1.0,
        }),
        dict({
          'condition': 'partlycloudy',
          'datetime': '2022-10-08 15:00:00+00:00',
          'precipitation_probability': 0,
          'temperature': 10.0,
          'templow': 3.0,
        }),
      ]),
    }),
  })
# ---
# name: test_forecast_daily_with_some_previous_days_data
  dict({
    'weather.home_forecast': dict({
      'forecast': list([
        dict({
          'condition': 'clear-night',
          'datetime': '2022-10-03 15:00:00+00:00',
          'precipitation_probability': 0,
          'temperature': None,
          'templow': -1.0,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2022-10-04 15:00:00+00:00',
          'precipitation_probability': 0,
          'temperature': 18.0,
          'templow': 3.0,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2022-10-05 15:00:00+00:00',
          'precipitation_probability': 0,
          'temperature': 20.0,
          'templow': 9.0,
        }),
        dict({
          'condition': 'partlycloudy',
          'datetime': '2022-10-06 15:00:00+00:00',
          'precipitation_probability': 0,
          'temperature': 20.0,
          'templow': 7.0,
        }),
        dict({
          'condition': 'rainy',
          'datetime': '2022-10-07 15:00:00+00:00',
          'precipitation_probability': 40,
          'temperature': 13.0,
          'templow': 1.0,
        }),
        dict({
          'condition': 'partlycloudy',
          'datetime': '2022-10-08 15:00:00+00:00',
          'precipitation_probability': 0,
          'temperature': 10.0,
          'templow': 3.0,
        }),
      ]),
    }),
  })
# ---
